### ISSUE TITLE
TORCH_NCCL_ENABLE_TIMING break nccl/matmul overlapping

### ISSUE THREAD

**Issue by cos120** [pytorch/pytorch#143890]
### 🐛 Describe the bug

I am using Megatron-LM for training, I found that if I set `TORCH_NCCL_ENABLE_TIMING=1`, all overlaping kernels in Megatron-LM will not overlapped, including dw/dx backward in layer norm and zero1 reduce scatter/allgather not overlapping with matmul.

I have submit a issue to `TransformerEngine`
https://github.com/NVIDIA/TransformerEngine/issues/1353, maybe it relates to `CUDA_DEVICE_MAX_CONNECTIONS=1`

### Versions

I am using 4 A100-SXM4 with pytorch2.4.0+cu124 and mcore0.9.0 with transformer engine(0.11.0+fc03478)

### update1

I use fsdp1 with 8 A100-SXM4 with pytorch2.4.0+cu124, and I tracing 3 configurations
- TORCH_NCCL_ENABLE_TIMING=1
not break overlapping of NCCL and matmul
- CUDA_DEVICE_MAX_CONNECTIONS=1
break reduce scatter in `FullyShardedDataParallel._post_backward_hook`
![image](https://github.com/user-attachments/assets/42f3fae0-9554-400b-8ef0-ca0b99541c4c)

- CUDA_DEVICE_MAX_CONNECTIONS=1 TORCH_NCCL_ENABLE_TIMING=1
break all overlap, including allgather in forward
![forward allgather overlap](https://github.com/user-attachments/assets/a608c054-8c64-4a0d-9f7f-198fe145b03e)
![backward allgather/reduce scatter](https://github.com/user-attachments/assets/495b07a1-99c2-47ef-90d1-a16b386894b4)

I upload timeline and reproduce code, just set different env and run 
```bash
CUDA_DEVICE_MAX_CONNECTIONS=1 TORCH_NCCL_ENABLE_TIMING=1 python -m torch.distributed.run --master-addr localhost --master-port 5555 --nnodes 1 --nproc-per-node 8 --node-rank 0 
```

[timeline.tar.gz](https://github.com/user-attachments/files/18326850/timeline.tar.gz)

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

**Comment by awgu** [comment]
Can @fduwjj @wconstab @kwen2501 or anyone familiar with `TORCH_NCCL_ENABLE_TIMING` take a look at this? `CUDA_DEVICE_MAX_CONNECTIONS=1` should prevent stream-stream overlapping.

**Comment by wconstab** [comment]
I'm pretty surprised to hear that enabling timing changes overlapping.

Its not clear if you also changed the setting of `CUDA_DEVICE_MAX_CONNECTIONS=1`?  I'm not familiar with this env.

**Comment by kwen2501** [comment]
CUDA_DEVICE_MAX_CONNECTION controls the numbers of hardware queues CUDA streams map onto. If it is set to 1, it will serialize all streams. This [post](https://forums.developer.nvidia.com/t/cuda-device-max-connections-and-pci-e-traffic/262962) has a nice and detailed explanation.

Megatron sets it to 1 after sequence parallel to manually control the execution order between the comm kernel and the GEMM kernel, per the [discussion](https://github.com/NVIDIA/Megatron-LM/issues/533) here.

**Comment by cos120** [comment]
@awgu @wconstab @kwen2501 I update timeline and reproduce code, check please, thanks!

**Comment by cos120** [comment]
> CUDA_DEVICE_MAX_CONNECTION controls the numbers of hardware queues CUDA streams map onto. If it is set to 1, it will serialize all streams. This [post](https://forums.developer.nvidia.com/t/cuda-device-max-connections-and-pci-e-traffic/262962) has a nice and detailed explanation.
> 
> Megatron sets it to 1 after sequence parallel to manually control the execution order between the comm kernel and the GEMM kernel, per the [discussion](https://github.com/NVIDIA/Megatron-LM/issues/533) here.

I check `TORCH_NCCL_ENABLE_TIMING` just add `cudaEventRecord` for NCCL kernels, I do not know why `CUDA_DEVICE_MAX_CONNECTIONS=1` will break overlap. As I known `CUDA_DEVICE_MAX_CONNECTIONS=1` only ensure the kernel scheduling is same as cpu issue kernel as FIFO, A100 has 108 sm, NCCL only use 24, it's enough for matmul to launch.

**Comment by kwen2501** [comment]
> As I known CUDA_DEVICE_MAX_CONNECTIONS=1 only ensure the kernel scheduling is same as cpu issue kernel as FIFO

Hmm, that's a good point. 

That said, I have no idea how many SMs a matmul can maximally request. But I've seen cases where matmul and comm blocks each other from launching.

**Comment by timmoon10** [comment]
I reproduce @cos120's observation that `TORCH_NCCL_ENABLE_TIMING=1` disrupts overlapping. Here is a pure PyTorch script that overlaps all-gather with an entrywise kernel:
<details> <summary> pyt_timing_test.py </summary>

```python
import functools
import os

import torch


@functools.cache
def world_group() -> torch.distributed.ProcessGroup:
    """Get NCCL process group, initializing if needed"""
    world_size = int(os.environ["WORLD_SIZE"])
    rank = int(os.environ["LOCAL_RANK"])
    torch.cuda.set_device(rank)
    group = torch.distributed.init_process_group(
        "nccl",
        init_method="file:///tmp/rdzv",
        world_size=world_size,
        rank=rank,
    )
    return group


def main() -> None:

    # Parallel config
    group = world_group()
    group_size = torch.distributed.get_world_size(group)
    rank = torch.distributed.get_rank(group)

    # Buffer for communication
    all_gather_size = 2 ** 30
    all_gather_buffer = torch.zeros(
        all_gather_size * group_size,
        dtype=torch.uint8,
        device="cuda",
    )
    all_gather_buffer_local = all_gather_buffer[rank * all_gather_size : (rank+1) * all_gather_size]

    # Buffer for compute
    compute_size = 2 ** 30
    compute_buffer = torch.ones(compute_size, dtype=torch.float32, device="cuda")

    # Overlap communication and computation
    default_stream = torch.cuda.default_stream()
    comm_stream = torch.cuda.Stream()
    for _ in range(10):
        comm_stream.wait_stream(default_stream)
        default_stream.wait_stream(comm_stream)
        with torch.cuda.stream(comm_stream):
            torch.distributed.all_gather_into_tensor(
                all_gather_buffer,
                all_gather_buffer_local,
                group=group,
            )
        torch.cos(compute_buffer)
    torch.cuda.synchronize()

    # Print status
    print(f"{rank=}: finished...")

if __name__ == "__main__":
    main()
```

</details>

You can run with `torchrun --nproc_per_node 4 pyt_timing_test.py`.

Capturing an nsys profile with 4 L40s, I see that `vectorized_elementwise_kernel` overlaps as expected with `ncclDevKernel_AllGather_RING_LL`:

![image](https://github.com/user-attachments/assets/0bbdef76-a371-4179-b887-05b0843263c5)

However, when I set `TORCH_NCCL_ENABLE_TIMING=1`, the kernels are serialized:

![image](https://github.com/user-attachments/assets/52ce5887-4888-4a31-a98c-d180680ce173)

I am running with PyTorch 2.6.0 and CUDA 12.6.

**Comment by wconstab** [comment]
still not too sure what's going on, but I wanted to point out that 'comm_stream' is a bit confusing since c10d is actually managing its own stream under the hood.  I wonder if you'd get different results if you use all_Gather_into_tensor(async_op=True) and use the resulting 'work.wait()` to sync the implicit c10d stream back to your default stream rather than managing your own comm stream.  

In any case, both ways should work.  I also wonder if `TORCH_NCCL_CUDA_EVENT_CACHE` env makes a difference.  In the event cache, iirc, we are separating the events by device but i don't think we care which stream they were last used with when we reuse them.

**Comment by fduwjj** [comment]
@wconstab No the stream information is unknown to the cache.

**Comment by wconstab** [comment]
Oddly enough, I did see some difference when I repro locally.  I was lazy about installing nsys but I just used torch profiler with a slight tweak to the orig repro script.
[new repro script](https://gist.github.com/wconstab/462c28b9e9d034ccb399148058dd3638)

when enable_timing=1 i see perfect overlap
<img width="2074" alt="image" src="https://github.com/user-attachments/assets/11418fbd-025a-421f-866b-743b28913188" />

with enable_timing = 0, i see less perfect overlap.  (but still some overlap)
<img width="2089" alt="image" src="https://github.com/user-attachments/assets/13650a5b-4af1-4af4-89eb-571be2c75d1b" />

**Comment by c-p-i-o** [comment]
From my attempts:

CUDA_DEVICE_MAX_CONNECTIONS=1
TORCH_NCCL_ENABLE_TIMING=1
small gaps



CUDA_DEVICE_MAX_CONNECTIONS=0
TORCH_NCCL_ENABLE_TIMING=1
no gaps

![Screenshot 2025-01-08 at 7 35 56 PM](https://github.com/user-attachments/assets/812d3165-6f83-4a52-8d72-445befce318d)
![Screenshot 2025-01-08 at 7 36 27 PM](https://github.com/user-attachments/assets/285c258c-0420-4040-a867-bc14393cfab3)

**Comment by kwen2501** [comment]
@timmoon10 @wconstab thanks for the repro. Can you please clarify if your experiment has any setting for `CUDA_DEVICE_MAX_CONNECTIONS`?

@c-p-i-o "small gap" means no overlap, it seem? I also don't know what effect `CUDA_DEVICE_MAX_CONNECTIONS=0` will create -- is 0 a valid value?

**Comment by kwen2501** [comment]
Hi @timmoon10, I took a deeper look in your repro, can you please share why the two streams were made to wait for each other?
```
comm_stream.wait_stream(default_stream)
default_stream.wait_stream(comm_stream)
```

As @wconstab mentioned, a common way to achieve comm-compute overlap is adding the `async_op=True` flag:
```
torch.distributed.all_gather_into_tensor(
  all_gather_buffer,
  all_gather_buffer_local,
  group=group,
  async_op=True,
)
torch.cos(compute_buffer)
```

**Comment by kwen2501** [comment]
Back to @cos120 's "update 1", it seems `TORCH_NCCL_ENABLE_TIMING=1` alone would *not* break overlap. 

The issue manifests only when `CUDA_DEVICE_MAX_CONNECTIONS=1`. But it is still an issue as long as Megatron has this setting, because PyTorch 2.6 started to default `ENABLE_TIMING` to 1. @fduwjj @c-p-i-o right?

**Comment by wconstab** [comment]
Edited my original post: I didn't specify `CUDA_DEVICE_MAX_CONNECTIONS` so I used the default value.

Is there documentation for what `CUDA_DEVICE_MAX_CONNECTIONS` does and how it is supposed to affect overlapping?

**Comment by c-p-i-o** [comment]
> The issue manifests only when `CUDA_DEVICE_MAX_CONNECTIONS=1`. But it is still an issue as long as Megatron has this setting, because PyTorch 2.6 started to default `ENABLE_TIMING` to 1. @fduwjj @c-p-i-o right?

@kwen2501 , we haven't changed PyTorch for `ENABLE_TIMING` to 1 in 2.6 [code](https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L934)

The default still is that TIMING isn't enabled. So a user would currently have to enable it explicitly.
That said, we're testing this switch out at Meta and rolling it out internally. This experiment is still in its infancy.

**Comment by timmoon10** [comment]
Thanks all for helping to look into this.

### `CUDA_DEVICE_MAX_CONNECTIONS`

`CUDA_DEVICE_MAX_CONNECTIONS` controls the number of work queues that the scheduler uses to send CUDA kernels to the GPU. By running with `CUDA_DEVICE_MAX_CONNECTIONS=1`, we are forcing the kernels to _start_ in the order they were launched on CPU. We do this in Megatron-LM to avoid some race conditions that can prevent overlapping (e.g. GEMM might monopolize SMs and prevent a small NCCL kernel from starting, so we want NCCL to start first, claim a few SMs, and let GEMM take the rest). 

### Answering questions

- I did all my runs with `CUDA_DEVICE_MAX_CONNECTIONS=1`. This problem definitely seems specific to this case.
- The streams syncs are there to mimic a tensor-parallel linear layer, which might overlap a GEMM and some tensor-parallel communication, and then synchronize before moving to the next layer.
- We could accomplish the desired behavior in this repro by launching NCCL with `async_op=True`, and it would probably remove some redundant stream synchronizations and overhead. However, it shouldn't affect anything logically so it would be good to understand what's going on before making drastic changes.

### Rambling thoughts

As far as I can tell, the only difference with `TORCH_NCCL_ENABLE_TIMING=1` is an extra `cudaEventRecord` before the NCCL call. As I see it:
```c++
if (TORCH_NCCL_ENABLE_TIMING) {
  cudaEventRecord();
}
ncclAllGather_kernel<<<..., stream0>>>(...);
// Miscellaneous CUDA event logic
cos_kernel1<<<..., stream1>>>(...);
cos_kernel2<<<..., stream1>>>(...);
```
I could see how `TORCH_NCCL_ENABLE_TIMING=1` could interfere with overlapping if it changed the behavior between the communication and compute kernels, but the change seems to happen in a completely separate place.

**Comment by cos120** [comment]
Sorry for later replay, thanks for all guys!

I have found something in FSDP, I think it relates with `CUDA_DEVICE_MAX_CONNECTIONS`.

For my FSDP training code, I found that the launching `rms_norm` on host is **after** `ncclDevKernel_ReduceScatter_Sum_bf16_RING_LL`, this is pure cpu code, not matter how `CUDA_DEVICE_MAX_CONNECTIONS` or `TORCH_NCCL_ENABLE_TIMING` is setting.

 But in timeline, the kernel `rms_norm` is scheduled **before** `ncclDevKernel_ReduceScatter_Sum_bf16_RING_LL`

And I do not set `CUDA_DEVICE_MAX_CONNECTIONS`, the kernels are overlapping.

![Image](https://github.com/user-attachments/assets/b8aa1789-4ab4-4c04-ba49-4a98113ab91f)

![Image](https://github.com/user-attachments/assets/022eaea9-296a-4f78-981e-c5ceb44c3c6b)

But with `CUDA_DEVICE_MAX_CONNECTIONS=1`, since `rms_norm` is **after** `ncclDevKernel_ReduceScatter_Sum_bf16_RING_LL` on host, cuda will schedule kernel `rms_norm`  **after** `ncclDevKernel_ReduceScatter_Sum_bf16_RING_LL`, this is right totally!

But with `CUDA_DEVICE_MAX_CONNECTIONS=1 TORCH_NCCL_ENABLE_TIMING=1`, all kernels are serialize. I am confusing that `TORCH_NCCL_ENABLE_TIMING=1` will let event create with `cudaEventDefault`, maybe `CUDA_DEVICE_MAX_CONNECTIONS=1` will take affects on event with `cudaEventDefault` but not affects on event with `cudaEventDisableTiming`?

**Comment by c-p-i-o** [comment]
@cos120 - do you still need help with this issue?

**Comment by wconstab** [comment]
@cos120 it wasn't entirely clear from your post on Jan13 if the entire issue is resolved after your discovery of FSDP-specific behavior relating to `CUDA_DEVICE_MAX_CONNECTIONS`.  We are still planning to land the `TORCH_NCCL_ENABLE_TIMING=1` default based on the conclusion that it is not affecting comm overlapping.  Please confirm.  I will close this issue for now, reopen if there is still a lingering problem.  Thanks!

also cc @timmoon10 @eqy @fduwjj

**Comment by cos120** [comment]
> [@cos120](https://github.com/cos120) - do you still need help with this issue?

Thanks for your helping! I am clear with this issue.

**Comment by cos120** [comment]
> [@cos120](https://github.com/cos120) it wasn't entirely clear from your post on Jan13 if the entire issue is resolved after your discovery of FSDP-specific behavior relating to `CUDA_DEVICE_MAX_CONNECTIONS`. We are still planning to land the `TORCH_NCCL_ENABLE_TIMING=1` default based on the conclusion that it is not affecting comm overlapping. Please confirm. I will close this issue for now, reopen if there is still a lingering problem. Thanks!
> 
> also cc [@timmoon10](https://github.com/timmoon10) [@eqy](https://github.com/eqy) [@fduwjj](https://github.com/fduwjj)

`CUDA_DEVICE_MAX_CONNECTIONS=1` is coming from Megatron-LM,  if `TORCH_NCCL_ENABLE_TIMING=1` will be default env, I think this will cause performance downgrade in Megatron-LM, but not FSDP.
